
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../plotting/">
      
      
        <link rel="next" href="../reference/capcruncher/api/annotate/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.4">
    
    
      
        <title>CLI Reference - CapCruncher Documentation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.8608ea7d.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#cli-reference" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="CapCruncher Documentation" class="md-header__button md-logo" aria-label="CapCruncher Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            CapCruncher Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              CLI Reference
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/sims-lab/CapCruncher" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="CapCruncher Documentation" class="md-nav__button md-logo" aria-label="CapCruncher Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    CapCruncher Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/sims-lab/CapCruncher" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../pipeline/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pipeline
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../cluster_config/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Cluster Setup
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../tips/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Hints and Tips
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../plotting/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Plotting CapCruncher output
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    CLI Reference
    
  </span>
  

      </a>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
        
          
          <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    API Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8">
            <span class="md-nav__icon md-icon"></span>
            API Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8_1" >
        
          
          <label class="md-nav__link" for="__nav_8_1" id="__nav_8_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    capcruncher
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_8_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8_1">
            <span class="md-nav__icon md-icon"></span>
            capcruncher
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8_1_1" >
        
          
          <label class="md-nav__link" for="__nav_8_1_1" id="__nav_8_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    api
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_8_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8_1_1">
            <span class="md-nav__icon md-icon"></span>
            api
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/capcruncher/api/annotate/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    annotate
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/capcruncher/api/deduplicate/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    deduplicate
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/capcruncher/api/filter/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    filter
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/capcruncher/api/io/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    io
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/capcruncher/api/pileup/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    pileup
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/capcruncher/api/plotting/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    plotting
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/capcruncher/api/statistics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    statistics
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/capcruncher/api/storage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    storage
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="cli-reference">CLI Reference<a class="headerlink" href="#cli-reference" title="Permanent link">&para;</a></h1>
<p>This page provides documentation for CapCruncher command line tools.</p>
<h1 id="capcruncher">capcruncher<a class="headerlink" href="#capcruncher" title="Permanent link">&para;</a></h1>
<p>An end to end solution for processing: Capture-C, Tri-C and Tiled-C data.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-0-1">capcruncher [OPTIONS] COMMAND [ARGS]...
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-1-1">  --version  Show the version and exit.
</span><span id="__span-1-2">  --help     Show this message and exit.
</span></code></pre></div>
<h2 id="alignments">alignments<a class="headerlink" href="#alignments" title="Permanent link">&para;</a></h2>
<p>Alignment annotation, identification and deduplication.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-2-1">capcruncher alignments [OPTIONS] COMMAND [ARGS]...
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-3-1">  --help  Show this message and exit.
</span></code></pre></div>
<h3 id="annotate">annotate<a class="headerlink" href="#annotate" title="Permanent link">&para;</a></h3>
<p>Annotates a bed file with other bed files using bedtools intersect.</p>
<p>Whilst bedtools intersect allows for interval names and counts to be used for annotating intervals, this command
provides the ability to annotate intervals with both interval names and counts at the same time. As the pipeline allows
for empty bed files, this command has built in support to deal with blank/malformed bed files and will return default N/A values.</p>
<p>Prior to interval annotation, the bed file to be intersected is validated and duplicate entries/multimapping reads are removed
to ensure consistent annotations and prevent issues with reporter identification.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-4-1">capcruncher alignments annotate [OPTIONS] SLICES
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-5-1">  -a, --actions [get|count]       Determines if the overlaps are counted or if
</span><span id="__span-5-2">                                  the name should just be reported
</span><span id="__span-5-3">  -b, --bed_files TEXT            Bed file(s) to intersect with slices
</span><span id="__span-5-4">  -n, --names TEXT                Names to use as column names for the output
</span><span id="__span-5-5">                                  tsv file.
</span><span id="__span-5-6">  -f, --overlap_fractions FLOAT   The minimum overlap required for an
</span><span id="__span-5-7">                                  intersection between two intervals to be
</span><span id="__span-5-8">                                  reported.
</span><span id="__span-5-9">  -t, --dtypes TEXT               Data type for column
</span><span id="__span-5-10">  -o, --output TEXT               Path for the annotated slices to be output.
</span><span id="__span-5-11">  --duplicates [remove]           Method to use for reconciling duplicate
</span><span id="__span-5-12">                                  slices (i.e. multimapping). Currently only
</span><span id="__span-5-13">                                  &#39;remove&#39; is supported.
</span><span id="__span-5-14">  -p, --n_cores INTEGER           Intersections are performed in parallel, set
</span><span id="__span-5-15">                                  this to the number of intersections required
</span><span id="__span-5-16">  --invalid_bed_action [ignore|error]
</span><span id="__span-5-17">                                  Method to deal with invalid bed files e.g.
</span><span id="__span-5-18">                                  blank or incorrectly formatted. Setting this
</span><span id="__span-5-19">                                  to &#39;ignore&#39; will report default N/A values
</span><span id="__span-5-20">                                  (either &#39;.&#39; or 0) for invalid files
</span><span id="__span-5-21">  --blacklist TEXT                Regions to remove from the BAM file prior to
</span><span id="__span-5-22">                                  annotation
</span><span id="__span-5-23">  --prioritize-cis-slices         Attempts to prevent slices on the most
</span><span id="__span-5-24">                                  common chromosome in a fragment (ideally cis
</span><span id="__span-5-25">                                  to the viewpoint) being removed by
</span><span id="__span-5-26">                                  deduplication
</span><span id="__span-5-27">  --priority-chroms TEXT          A comma separated list of chromosomes to
</span><span id="__span-5-28">                                  prioritize during deduplication
</span><span id="__span-5-29">  --help                          Show this message and exit.
</span></code></pre></div>
<h3 id="filter">filter<a class="headerlink" href="#filter" title="Permanent link">&para;</a></h3>
<p>Removes unwanted aligned slices and identifies reporters.</p>
<p>Parses a BAM file and merges this with a supplied annotation to identify unwanted slices.
Filtering can be tuned for Capture-C, Tri-C and Tiled-C data to ensure optimal filtering.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-6-1">capcruncher alignments filter [OPTIONS] {capture|tri|tiled}
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-7-1">  -b, --bam TEXT                Bam file to process  [required]
</span><span id="__span-7-2">  -a, --annotations TEXT        Annotations for the bam file that must contain
</span><span id="__span-7-3">                                the required columns, see description.
</span><span id="__span-7-4">                                [required]
</span><span id="__span-7-5">  --custom-filtering TEXT       Custom filtering to be used. This must be
</span><span id="__span-7-6">                                supplied as a path to a yaml file.
</span><span id="__span-7-7">  -o, --output_prefix TEXT      Output prefix for deduplicated fastq file(s)
</span><span id="__span-7-8">  --statistics TEXT             Output path for stats file
</span><span id="__span-7-9">  --sample-name TEXT            Name of sample e.g. DOX_treated_1
</span><span id="__span-7-10">  --read-type [flashed|pe]      Type of read
</span><span id="__span-7-11">  --fragments / --no-fragments  Determines if read fragment aggregations are
</span><span id="__span-7-12">                                produced
</span><span id="__span-7-13">  --help                        Show this message and exit.
</span></code></pre></div>
<h2 id="fastq">fastq<a class="headerlink" href="#fastq" title="Permanent link">&para;</a></h2>
<p>Fastq splitting, deduplication and digestion.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-8-1">capcruncher fastq [OPTIONS] COMMAND [ARGS]...
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-9-1">  --help  Show this message and exit.
</span></code></pre></div>
<h3 id="deduplicate">deduplicate<a class="headerlink" href="#deduplicate" title="Permanent link">&para;</a></h3>
<p>Identifies PCR duplicate fragments from Fastq files.</p>
<p>PCR duplicates are very commonly present in Capture-C/Tri-C/Tiled-C data and must be removed
for accurate analysis. These commands attempt to identify and remove duplicate reads/fragments
from fastq file(s) to speed up downstream analysis.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-10-1">capcruncher fastq deduplicate [OPTIONS]
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-11-1">  -1, --fastq1 TEXT         Read 1 FASTQ files  [required]
</span><span id="__span-11-2">  -2, --fastq2 TEXT         Read 2 FASTQ files  [required]
</span><span id="__span-11-3">  -o, --output-prefix TEXT  Output prefix for deduplicated FASTQ files
</span><span id="__span-11-4">  --sample-name TEXT        Name of sample e.g. DOX_treated_1
</span><span id="__span-11-5">  -s, --statistics TEXT     Statistics output file name
</span><span id="__span-11-6">  --shuffle                 Shuffle reads before deduplication
</span><span id="__span-11-7">  --help                    Show this message and exit.
</span></code></pre></div>
<h3 id="digest">digest<a class="headerlink" href="#digest" title="Permanent link">&para;</a></h3>
<p>Performs in silico digestion of one or a pair of fastq files.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-12-1">capcruncher fastq digest [OPTIONS] FASTQS...
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-13-1">  -r, --restriction_enzyme TEXT   Restriction enzyme name or sequence to use
</span><span id="__span-13-2">                                  for in silico digestion.  [required]
</span><span id="__span-13-3">  -m, --mode [flashed|pe]         Digestion mode. Combined (Flashed) or non-
</span><span id="__span-13-4">                                  combined (PE) read pairs.  [required]
</span><span id="__span-13-5">  -o, --output_file TEXT
</span><span id="__span-13-6">  --minimum_slice_length INTEGER
</span><span id="__span-13-7">  --statistics TEXT               Output path for stats file
</span><span id="__span-13-8">  --sample-name TEXT              Name of sample e.g. DOX_treated_1. Required
</span><span id="__span-13-9">                                  for correct statistics.
</span><span id="__span-13-10">  --help                          Show this message and exit.
</span></code></pre></div>
<h3 id="split">split<a class="headerlink" href="#split" title="Permanent link">&para;</a></h3>
<p>Splits fastq file(s) into equal chunks of n reads.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-14-1">capcruncher fastq split [OPTIONS] INPUT_FILES...
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-15-1">  -m, --method [python|unix]   Method to use for splitting
</span><span id="__span-15-2">  -o, --output_prefix TEXT     Output prefix for deduplicated fastq file(s)
</span><span id="__span-15-3">  --compression_level INTEGER  Level of compression for output files
</span><span id="__span-15-4">  -n, --n_reads INTEGER        Number of reads per fastq file
</span><span id="__span-15-5">  --gzip / --no-gzip           Determines if files are gziped or not
</span><span id="__span-15-6">  -p, --n_cores INTEGER
</span><span id="__span-15-7">  -s, --suffix TEXT            Suffix to add to output files (ignore
</span><span id="__span-15-8">                               {read_number}.fastq as this is added
</span><span id="__span-15-9">                               automatically)
</span><span id="__span-15-10">  --help                       Show this message and exit.
</span></code></pre></div>
<h2 id="genome">genome<a class="headerlink" href="#genome" title="Permanent link">&para;</a></h2>
<p>Genome wide methods digestion.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-16-1">capcruncher genome [OPTIONS] COMMAND [ARGS]...
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-17-1">  --help  Show this message and exit.
</span></code></pre></div>
<h3 id="digest_1">digest<a class="headerlink" href="#digest_1" title="Permanent link">&para;</a></h3>
<p>Performs in silico digestion of a genome in fasta format.</p>
<p>Digests the supplied genome fasta file and generates a bed file containing the
locations of all restriction fragments produced by the supplied restriction enzyme.</p>
<p>A log file recording the number of restriction fragments for the suplied genome is also
generated.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-18-1">capcruncher genome digest [OPTIONS] INPUT_FASTA
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-19-1">  -r, --recognition_site TEXT  Recognition enzyme or sequence  [required]
</span><span id="__span-19-2">  -l, --logfile TEXT           Path for digestion log file
</span><span id="__span-19-3">  -o, --output_file TEXT       Output file path
</span><span id="__span-19-4">  --remove_cutsite BOOLEAN     Exclude the recognition sequence from the
</span><span id="__span-19-5">                               output
</span><span id="__span-19-6">  --sort                       Sorts the output bed file by chromosome and
</span><span id="__span-19-7">                               start coord.
</span><span id="__span-19-8">  --help                       Show this message and exit.
</span></code></pre></div>
<h2 id="interactions">interactions<a class="headerlink" href="#interactions" title="Permanent link">&para;</a></h2>
<p>Reporter counting, storing, comparison and pileups</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-20-1">capcruncher interactions [OPTIONS] COMMAND [ARGS]...
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-21-1">  --help  Show this message and exit.
</span></code></pre></div>
<h3 id="compare">compare<a class="headerlink" href="#compare" title="Permanent link">&para;</a></h3>
<p>Compare bedgraphs and CapCruncher cooler files.</p>
<p>These commands allow for specific viewpoints to be extracted from CapCruncher HDF5 files and perform:</p>
<div class="language-text highlight"><pre><span></span><code>1. User defined groupby aggregations.

2. Comparisons between conditions.

3. Identification of differential interactions between conditions.
</code></pre></div>
<p>See subcommands for details.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-22-1">capcruncher interactions compare [OPTIONS] COMMAND [ARGS]...
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-23-1">  --help  Show this message and exit.
</span></code></pre></div>
<h4 id="concat">concat<a class="headerlink" href="#concat" title="Permanent link">&para;</a></h4>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-24-1">capcruncher interactions compare concat [OPTIONS] INFILES...
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-25-1">  -f, --format [auto|bedgraph|cooler]
</span><span id="__span-25-2">                                  Input file format
</span><span id="__span-25-3">  -o, --output TEXT               Output file name
</span><span id="__span-25-4">  -v, --viewpoint TEXT            Viewpoint to extract
</span><span id="__span-25-5">  -r, --resolution TEXT           Resolution to extract
</span><span id="__span-25-6">  --region TEXT                   Limit to specific coordinates in the format
</span><span id="__span-25-7">                                  chrom:start-end
</span><span id="__span-25-8">  --normalisation [raw|n_cis|region]
</span><span id="__span-25-9">                                  Method to use interaction normalisation
</span><span id="__span-25-10">  --normalisation-regions TEXT    Regions to use for interaction
</span><span id="__span-25-11">                                  normalisation. The --normalisation method
</span><span id="__span-25-12">                                  MUST be &#39;region&#39;
</span><span id="__span-25-13">  --scale_factor INTEGER          Scale factor to use for bedgraph
</span><span id="__span-25-14">                                  normalisation
</span><span id="__span-25-15">  -p, --n_cores INTEGER           Number of cores to use for extracting
</span><span id="__span-25-16">                                  bedgraphs
</span><span id="__span-25-17">  --help                          Show this message and exit.
</span></code></pre></div>
<h4 id="differential">differential<a class="headerlink" href="#differential" title="Permanent link">&para;</a></h4>
<p>Perform differential testing on CapCruncher HDF5 files.</p>
<p>This command performs differential testing on CapCruncher HDF5 files. It requires a design matrix
and a contrast to test. The design matrix should be a tab separated file with the first column
containing the sample names and the remaining columns containing the conditions. The contrast
should specify the name of the column in the design matrix to test. The output is a tab separated bedgraph.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-26-1">capcruncher interactions compare differential [OPTIONS] INTERACTION_FILES...
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-27-1">  -o, --output-prefix TEXT        Output file prefix
</span><span id="__span-27-2">  -v, --viewpoint TEXT            Viewpoint to extract  [required]
</span><span id="__span-27-3">  -d, --design-matrix TEXT        Design matrix file  [required]
</span><span id="__span-27-4">  -c, --contrast TEXT             Contrast to test
</span><span id="__span-27-5">  -r, --regions-of-interest TEXT  Regions of interest to test for differential
</span><span id="__span-27-6">                                  interactions
</span><span id="__span-27-7">  --viewpoint-distance INTEGER    Distance from viewpoint to test for
</span><span id="__span-27-8">                                  differential interactions
</span><span id="__span-27-9">  --threshold-count INTEGER       Minimum number of interactions to test for
</span><span id="__span-27-10">                                  differential interactions
</span><span id="__span-27-11">  --threshold-q FLOAT             Minimum q-value to test for differential
</span><span id="__span-27-12">                                  interactions
</span><span id="__span-27-13">  --help                          Show this message and exit.
</span></code></pre></div>
<h4 id="summarise">summarise<a class="headerlink" href="#summarise" title="Permanent link">&para;</a></h4>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-28-1">capcruncher interactions compare summarise [OPTIONS] INFILE
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-29-1">  -d, --design-matrix TEXT        Design matrix file, should be formatted as a
</span><span id="__span-29-2">                                  tab separated file with the first column
</span><span id="__span-29-3">                                  containing the sample names and the other
</span><span id="__span-29-4">                                  column containing the conditions.
</span><span id="__span-29-5">  -o, --output-prefix TEXT        Output file prefix
</span><span id="__span-29-6">  -f, --output-format [bedgraph|tsv]
</span><span id="__span-29-7">  -m, --summary-methods TEXT      Summary methods to use for aggregation. Can
</span><span id="__span-29-8">                                  be any method in numpy or scipy.stats
</span><span id="__span-29-9">  -n, --group-names TEXT          Group names for aggregation
</span><span id="__span-29-10">  -c, --group-columns TEXT        Column names/numbers (0 indexed, the first
</span><span id="__span-29-11">                                  column after the end coordinate counts as 0)
</span><span id="__span-29-12">                                  for aggregation.
</span><span id="__span-29-13">  --subtraction                   Perform subtration between aggregated groups
</span><span id="__span-29-14">  --suffix TEXT                   Add a suffix before the file extension
</span><span id="__span-29-15">  --help                          Show this message and exit.
</span></code></pre></div>
<h3 id="count">count<a class="headerlink" href="#count" title="Permanent link">&para;</a></h3>
<p>Determines the number of captured restriction fragment interactions genome wide.</p>
<p>Counts the number of interactions between each restriction fragment and all other
restriction fragments in the fragment.</p>
<p>The output is a cooler formatted HDF5 file containing a single group containing
the interactions between restriction fragments.</p>
<p>See <code>https://cooler.readthedocs.io/en/latest/</code> for further details.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-30-1">capcruncher interactions count [OPTIONS] REPORTERS
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-31-1">  -o, --output TEXT            Name of output file
</span><span id="__span-31-2">  --remove_exclusions          Prevents analysis of fragments marked as
</span><span id="__span-31-3">                               proximity exclusions
</span><span id="__span-31-4">  --remove_capture             Prevents analysis of capture fragment
</span><span id="__span-31-5">                               interactions
</span><span id="__span-31-6">  --subsample FLOAT            Subsamples reporters before analysis of
</span><span id="__span-31-7">                               interactions
</span><span id="__span-31-8">  -f, --fragment-map TEXT      Path to digested genome bed file
</span><span id="__span-31-9">  -v, --viewpoint-path TEXT    Path to viewpoints file
</span><span id="__span-31-10">  -p, --n-cores INTEGER        Number of cores to use for counting.
</span><span id="__span-31-11">  --assay [capture|tri|tiled]
</span><span id="__span-31-12">  --help                       Show this message and exit.
</span></code></pre></div>
<h3 id="counts-to-cooler">counts-to-cooler<a class="headerlink" href="#counts-to-cooler" title="Permanent link">&para;</a></h3>
<p>Stores restriction fragment interaction combinations at the restriction fragment level.</p>
<p>Parses reporter restriction fragment interaction counts produced by
"capcruncher reporters count" and gerates a cooler formatted group in an HDF5 File.
See <code>https://cooler.readthedocs.io/en/latest/</code> for further details.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-32-1">capcruncher interactions counts-to-cooler [OPTIONS] COUNTS
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-33-1">  -f, --fragment-map TEXT    Path to digested genome bed file  [required]
</span><span id="__span-33-2">  -v, --viewpoint-path TEXT  Path to viewpoints file  [required]
</span><span id="__span-33-3">  -n, --viewpoint-name TEXT  Name of viewpoint to store
</span><span id="__span-33-4">  -g, --genome TEXT          Name of genome
</span><span id="__span-33-5">  --suffix TEXT              Suffix to append after the capture name for the
</span><span id="__span-33-6">                             output file
</span><span id="__span-33-7">  -o, --output TEXT          Name of output file. (Cooler formatted hdf5 file)
</span><span id="__span-33-8">  --help                     Show this message and exit.
</span></code></pre></div>
<h3 id="deduplicate_1">deduplicate<a class="headerlink" href="#deduplicate_1" title="Permanent link">&para;</a></h3>
<p>Identifies and removes duplicated aligned fragments.</p>
<p>PCR duplicates are very commonly present in Capture-C/Tri-C/Tiled-C data and must be removed
for accurate analysis. Unlike fastq deduplicate, this command removes fragments with identical
genomic coordinates.</p>
<p>Non-combined (pe) and combined (flashed) reads are treated slightly differently due to the increased
confidence that the ligation junction has been captured for the flashed reads.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-34-1">capcruncher interactions deduplicate [OPTIONS] SLICES
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-35-1">  -o, --output TEXT         Output prefix for directory of deduplicated slices
</span><span id="__span-35-2">  --statistics TEXT         Output prefix for stats file(s)
</span><span id="__span-35-3">  --sample-name TEXT        Name of sample e.g. DOX_treated_1
</span><span id="__span-35-4">  --read-type [flashed|pe]  Type of read
</span><span id="__span-35-5">  --help                    Show this message and exit.
</span></code></pre></div>
<h3 id="fragments-to-bins">fragments-to-bins<a class="headerlink" href="#fragments-to-bins" title="Permanent link">&para;</a></h3>
<p>Convert a cooler group containing restriction fragments to constant genomic windows</p>
<p>Parses a cooler group and aggregates restriction fragment interaction counts into
genomic bins of a specified size. If the normalise option is selected,
columns containing normalised counts are added to the pixels table of the output</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-36-1">capcruncher interactions fragments-to-bins [OPTIONS] COOLER_PATH
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-37-1">  -b, --binsizes INTEGER       Binsizes to use for windowing
</span><span id="__span-37-2">  --normalise                  Enables normalisation of interaction counts
</span><span id="__span-37-3">                               during windowing
</span><span id="__span-37-4">  --overlap_fraction FLOAT     Minimum overlap between genomic bins and
</span><span id="__span-37-5">                               restriction fragments for overlap
</span><span id="__span-37-6">  -p, --n_cores INTEGER        Number of cores used for binning
</span><span id="__span-37-7">  --scale-factor INTEGER       Scaling factor used for normalisation
</span><span id="__span-37-8">  --conversion_tables TEXT     Pickle file containing pre-computed fragment -&gt;
</span><span id="__span-37-9">                               bin conversions.
</span><span id="__span-37-10">  -o, --output TEXT            Name of output file. (Cooler formatted hdf5
</span><span id="__span-37-11">                               file)
</span><span id="__span-37-12">  --assay [capture|tri|tiled]
</span><span id="__span-37-13">  --help                       Show this message and exit.
</span></code></pre></div>
<h3 id="merge">merge<a class="headerlink" href="#merge" title="Permanent link">&para;</a></h3>
<p>Merges capcruncher HDF5 files together.</p>
<p>Produces a unified cooler with both restriction fragment and genomic bins whilst
reducing the storage space required by hard linking the "bins" tables to prevent duplication.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-38-1">capcruncher interactions merge [OPTIONS] COOLERS...
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-39-1">  -o, --output TEXT  Output file name
</span><span id="__span-39-2">  --help             Show this message and exit.
</span></code></pre></div>
<h3 id="pileup">pileup<a class="headerlink" href="#pileup" title="Permanent link">&para;</a></h3>
<p>Extracts reporters from a capture experiment and generates a bedgraph file.</p>
<p>Identifies reporters for a single probe (if a probe name is supplied) or all capture
probes present in a capture experiment HDF5 file.</p>
<p>The bedgraph generated can be normalised by the number of cis interactions for
inter experiment comparisons and/or extract pilups binned into even genomic windows.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-40-1">capcruncher interactions pileup [OPTIONS] URI
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-41-1">  -n, --viewpoint_names TEXT      Viewpoint to extract and convert to
</span><span id="__span-41-2">                                  bedgraph, if not provided will transform
</span><span id="__span-41-3">                                  all.
</span><span id="__span-41-4">  -o, --output_prefix TEXT        Output prefix for bedgraphs
</span><span id="__span-41-5">  --normalisation [raw|n_cis|region]
</span><span id="__span-41-6">                                  Method to use interaction normalisation
</span><span id="__span-41-7">  --normalisation-regions TEXT    Regions to use for interaction
</span><span id="__span-41-8">                                  normalisation. The --normalisation method
</span><span id="__span-41-9">                                  MUST be &#39;region&#39;
</span><span id="__span-41-10">  --binsize INTEGER               Binsize to use for converting bedgraph to
</span><span id="__span-41-11">                                  evenly sized genomic bins
</span><span id="__span-41-12">  --gzip                          Compress output using gzip
</span><span id="__span-41-13">  --scale-factor INTEGER          Scale factor to use for bedgraph
</span><span id="__span-41-14">                                  normalisation
</span><span id="__span-41-15">  --sparse / --dense              Produce bedgraph containing just positive
</span><span id="__span-41-16">                                  bins (sparse) or all bins (dense)
</span><span id="__span-41-17">  -f, --format [bedgraph|bigwig]  Output file format
</span><span id="__span-41-18">  --help                          Show this message and exit.
</span></code></pre></div>
<h2 id="pipeline">pipeline<a class="headerlink" href="#pipeline" title="Permanent link">&para;</a></h2>
<p>Runs the data processing pipeline</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-42-1">capcruncher pipeline [OPTIONS] [PIPELINE_OPTIONS]...
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-43-1">  -h, --help
</span><span id="__span-43-2">  --version
</span><span id="__span-43-3">  --logo / --no-logo  Show the capcruncher logo  [default: logo]
</span><span id="__span-43-4">  --version           Show the version and exit.
</span></code></pre></div>
<h2 id="pipeline-config">pipeline-config<a class="headerlink" href="#pipeline-config" title="Permanent link">&para;</a></h2>
<p>Configures the data processing pipeline</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-44-1">capcruncher pipeline-config [OPTIONS]
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-45-1">  -h, --help
</span><span id="__span-45-2">  --version
</span><span id="__span-45-3">  --version          Show the version and exit.
</span><span id="__span-45-4">  -i, --input PATH
</span><span id="__span-45-5">  --generate-design
</span></code></pre></div>
<h2 id="plot">plot<a class="headerlink" href="#plot" title="Permanent link">&para;</a></h2>
<p>Generates plots for the outputs produced by CapCruncher</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-46-1">capcruncher plot [OPTIONS]
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-47-1">  -r, --region TEXT    Genomic coordinates of the region to plot  [required]
</span><span id="__span-47-2">  -t, --template TEXT  TOML file containing the template for the plot
</span><span id="__span-47-3">                       [required]
</span><span id="__span-47-4">  -o, --output TEXT    Output file path. The file extension determines the
</span><span id="__span-47-5">                       output format.
</span><span id="__span-47-6">  --help               Show this message and exit.
</span></code></pre></div>
<h2 id="utilities">utilities<a class="headerlink" href="#utilities" title="Permanent link">&para;</a></h2>
<p>Contains miscellaneous functions</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-48-1">capcruncher utilities [OPTIONS] COMMAND [ARGS]...
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-49-1">  --help  Show this message and exit.
</span></code></pre></div>
<h3 id="cis-and-trans-stats">cis-and-trans-stats<a class="headerlink" href="#cis-and-trans-stats" title="Permanent link">&para;</a></h3>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-50-1">capcruncher utilities cis-and-trans-stats [OPTIONS] SLICES
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-51-1">  -o, --output TEXT            Output file name
</span><span id="__span-51-2">  --sample-name TEXT           Name of sample e.g. DOX_treated_1
</span><span id="__span-51-3">  --assay [capture|tri|tiled]  Assay used to generate slices
</span><span id="__span-51-4">  --help                       Show this message and exit.
</span></code></pre></div>
<h3 id="dump">dump<a class="headerlink" href="#dump" title="Permanent link">&para;</a></h3>
<p>Dumps the contents of a cooler or capcruncher parquet file to a TSV file</p>
<p>Args:
    path (str): Path to cooler or capcruncher parquet file
    viewpoint (str, optional): Viewpoint to extract. Defaults to None.
    resolution (int, optional): Resolution to extract. Only used for cooler (hdf5) files. Defaults to None.
    output (str, optional): Output file name. Defaults to "capcruncher_dump.tsv".</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-52-1">capcruncher utilities dump [OPTIONS] PATH
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-53-1">  -v, --viewpoint TEXT   Viewpoint to extract
</span><span id="__span-53-2">  -r, --resolution TEXT  Resolution to extract. Only used for cooler (hdf5)
</span><span id="__span-53-3">                         files
</span><span id="__span-53-4">  -o, --output TEXT      Output file name
</span><span id="__span-53-5">  --help                 Show this message and exit.
</span></code></pre></div>
<h3 id="gtf-to-bed12">gtf-to-bed12<a class="headerlink" href="#gtf-to-bed12" title="Permanent link">&para;</a></h3>
<p>Converts a GTF file to a BED12 file containing only 5' UTRs, 3' UTRs, and exons.</p>
<p>Args:
    gtf (str): Path to the input GTF file.
    output (str): Path to the output BED12 file.</p>
<p>Returns:
    None</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-54-1">capcruncher utilities gtf-to-bed12 [OPTIONS] GTF
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-55-1">  -o, --output TEXT  Output file name
</span><span id="__span-55-2">  --help             Show this message and exit.
</span></code></pre></div>
<h3 id="make-chicago-maps">make-chicago-maps<a class="headerlink" href="#make-chicago-maps" title="Permanent link">&para;</a></h3>
<p>Restriction map file (.rmap) - a bed file containing coordinates of the restriction fragments. By default, 4 columns: chr, start, end, fragmentID.
Bait map file (.baitmap) - a bed file containing coordinates of the baited restriction fragments, and their associated annotations. By default, 5 columns: chr, start, end, fragmentID, baitAnnotation. The regions specified in this file, including their fragmentIDs, must be an exact subset of those in the .rmap file. The baitAnnotation is a text field that is used only to annotate the output and plots.</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-56-1">capcruncher utilities make-chicago-maps [OPTIONS]
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-57-1">  --fragments TEXT      Path to fragments file (default: capcruncher_output/re
</span><span id="__span-57-2">                        sources/restriction_fragments/genome.digest.bed.gz)
</span><span id="__span-57-3">  --viewpoints TEXT     Path to viewpoints file used for capcruncher
</span><span id="__span-57-4">                        [required]
</span><span id="__span-57-5">  -o, --outputdir TEXT  Path to output directory  [required]
</span><span id="__span-57-6">  --help                Show this message and exit.
</span></code></pre></div>
<h3 id="regenerate-fastq">regenerate-fastq<a class="headerlink" href="#regenerate-fastq" title="Permanent link">&para;</a></h3>
<p>Regenerates a FASTQ file from a parquet file containing the required reads</p>
<p>Args:
    fastq1 (str): Path to the first FASTQ file
    fastq2 (str): Path to the second FASTQ file
    parquet_file (str, optional): Path to the parquet file from which to extract the required reads. Defaults to None.
    output (str, optional): Prefix for the output file. Defaults to "regenerated_".</p>
<p>Raises:
    AssertionError: If the specified parquet file does not exist.</p>
<p>Returns:
    None</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-58-1">capcruncher utilities regenerate-fastq [OPTIONS]
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-59-1">  -1, --fastq1 TEXT         Path to FASTQ file 1  [required]
</span><span id="__span-59-2">  -2, --fastq2 TEXT         Path to FASTQ file 2  [required]
</span><span id="__span-59-3">  -p, --parquet-file TEXT   Path to parquet file from which to extract the
</span><span id="__span-59-4">                            required reads  [required]
</span><span id="__span-59-5">  -o, --output-prefix TEXT  Output file prefix
</span><span id="__span-59-6">  --help                    Show this message and exit.
</span></code></pre></div>
<h3 id="viewpoint-coordinates">viewpoint-coordinates<a class="headerlink" href="#viewpoint-coordinates" title="Permanent link">&para;</a></h3>
<p>Aligns viewpoints to a genome and returns the coordinates of the viewpoint
in the genome.</p>
<p>Viewpoints can be supplied as a FASTA file or a TSV file with the first column
containing the name of the viewpoint and the second column containing the
sequence of the viewpoint.</p>
<p>Args:
    viewpoints (os.PathLike): Path to viewpoints
    genome (os.PathLike): Path to genome fasta file
    genome_indicies (os.PathLike, optional): Path to genome bowtie2 indices. Defaults to None.
    recognition_site (str, optional): Restriction site used. Defaults to "dpnii".
    output (os.PathLike, optional): Output file name. Defaults to "viewpoint_coordinates.bed".</p>
<p>Raises:
    ValueError: If viewpoints are not supplied in the correct format
    ValueError: If no bowtie2 indices are supplied</p>
<p><strong>Usage:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-60-1">capcruncher utilities viewpoint-coordinates [OPTIONS]
</span></code></pre></div>
<p><strong>Options:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-61-1">  -v, --viewpoints TEXT        Path to viewpoints  [required]
</span><span id="__span-61-2">  -g, --genome TEXT            Path to genome fasta file  [required]
</span><span id="__span-61-3">  -i, --genome-indicies TEXT   Path to genome bowtie2 indices  [required]
</span><span id="__span-61-4">  -r, --recognition-site TEXT  Restriction site used
</span><span id="__span-61-5">  -o, --output TEXT            Output file name
</span><span id="__span-61-6">  --help                       Show this message and exit.
</span></code></pre></div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["content.code.copy", "header.autohide"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.f1b6f286.min.js"></script>
      
    
  </body>
</html>